### ISSUE TITLE
[Performance] `tensordot` has substantial overhead

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#145731]
### 🐛 Describe the bug

`torch.tensordot(A, B)` is ~50% slower than `torch.dot(A.flatten(), B.flatten())` for matrices of size≤512 on GPU, and on CPU it is even slower than `torch.trace(A.T @ B)` until n=1024.

![Image](https://github.com/user-attachments/assets/970298ae-9bd8-45fd-b813-66d80921efc0)

![Image](https://github.com/user-attachments/assets/420a90ce-a16b-46e4-9100-ff113669bd51)

## Why this matters

A common mistake is to compute a quantity expressed in a math textbook / paper like $\text{trace}(A^T B)$ naively how it is written, which is $O(n^3)$, when it could be computed as $\langle A \mid B \rangle_{F} = \sum_{ij} A_{ij}B_{ij}$ which is $O(n^2)$. (The former computes all off-diagonal entries of the matrix product, which are irrelevant for the trace).

This would make a good lint rule, but the most natural implementation of the Frobenius inner product, which is `tensordot`, has this huge overhead and so should be currently avoided for small matrices.

## Expectation

- `tensordot(A, B)` should not be substantially slower than `torch.dot(A.flatten(), B.flatten())`.
- `tensordot(A, B)` should usually be faster than `torch.sum(A*B)`, as the latter cannot take advantage of fused multiply-add instructions.

## References:

- https://github.com/numpy/numpy/issues/25713
- https://github.com/astral-sh/ruff/issues/9664
- Code: https://gist.github.com/randolf-scholz/9b166ff6b112820d123037448369a0b7

### Versions

<details>

Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.31.2
Libc version: glibc-2.35

Python version: 3.12.8 (main, Dec  4 2024, 08:54:12) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-51-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 570.86.10
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
CPU family:                           6
Model:                                141
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          4800,0000
CPU min MHz:                          800,0000
BogoMIPS:                             4992.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             10 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.14.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.5.1
[pip3] torchinfo==1.8.0
[pip3] triton==3.1.0
[conda] Could not collect

</details>

cc @msaroufim @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

**Comment by nikitaved** [comment]
I confirm there is a regression. The code for `tensordot`, on the surface, looks fine. It does not seem to be doing cubic stuff. I suspect there might be some memory (layout?)-related issues:
```python
In [1]: import torch

In [2]: n = 512

In [3]: A = torch.rand(n, n, device="cuda")

In [4]: B = torch.rand(n, n, device="cuda")

In [5]: %timeit torch.tensordot(A, B, [[1, 0], [0, 1]]); torch.cuda.synchronize()
56.9 μs ± 29.9 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit torch.tensordot(A, B, [[1, 0], [1, 0]]); torch.cuda.synchronize()
47.5 μs ± 521 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [7]: %timeit torch.tensordot(A, B, [[0, 1], [0, 1]]); torch.cuda.synchronize()
29.7 μs ± 41.5 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: %timeit torch.tensordot(A, B, [[0, 1], [1, 0]]); torch.cuda.synchronize()
40.8 μs ± 68.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [9]: torch.tensordot(A, B, [[0, 1], [0, 1]])
Out[9]: tensor(65423.1953, device='cuda:0')

In [10]: (A * B).sum()
Out[10]: tensor(65423.1953, device='cuda:0')

In [11]: %timeit (A * B).sum(); torch.cuda.synchronize()
24.2 μs ± 210 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

```

**Comment by randolf-scholz** [comment]
It may just be overhead due to slow input parsing, see the linked numpy issue.

**Comment by nikitaved** [comment]
@randolf-scholz , I doubt that. Take a look at how `tensordot(A, B, [[0, 1], [0, 1]])` compares to `(A * B).sum()`.

**Comment by randolf-scholz** [comment]
Ah, ok. So likely it doesn't go over contiguous slices of memory due to layout. 

But even then, `tensordot(A, B, [[0, 1], [0, 1]])` is roughly 20% slower `(A * B).sum()`, which then is the reaming overhead from parsing I guess. 

Ideally, it should always be faster, since `(A * B).sum()` cannot take advantage of FMA instructions.

**Comment by randolf-scholz** [comment]
Well, in my experiments, `torch.dot(A.flatten(), B.flatten())` is always faster than `(A*B).sum()`, you can see there's a constant cap between the two for larger sizes.

**Comment by nikitaved** [comment]
> Well, in my experiments, `torch.dot(A.flatten(), B.flatten())` is always faster than `(A*B).sum()`, you can see there's a constant gap between the two for larger sizes (green and orange curves)

Yeah, that makes sense. We can easily dispatch to `dot` when reducing over all the dims, actually. That should be faster than `gemm` for sure.

**Comment by randolf-scholz** [comment]
One thing I have been wondering, since it also annoys me that often `einsum` is slower than equivalent formulation of the same contraction, is whether one could just optimize all of this nonsense away under the hood by having a kernel registry that for a given contraction selects the fastest available kernel.

For example, say we want to do `tensordot(A, B)` for 2 matrices.

1. Formulate as a general tensor contraction, in this case: `"ij,ij ->"`
2. canonicalize the signature into a hashable object `"ij,ij->"   =>  `((("v0", "v1"), ("v0", "v1")), ())`
3. select kernel for this contraction via hashmap lookup

I.e. using some large dictionary like the following as a jump table for tensor contraction kernels:

```python
type Arg = tuple[str, ...]
kernels: dict[tuple[tuple[Arg, ...], Arg]], Callable] = {
    # sum
    ((("v0",)), ()): torch.sum
    # vec-vec inner product
    ((("v0", "v0")), ()): torch.dot, 
    # vec-vec elementwise multiply
    ((("v0",), ("v0",)), ("v0",)): torch.multiply,
    # trace
    ((("v0", "v0")), ()): torch.trace,
    # matmul A B
    ((("v0", "v1"), ("v1", "v2")), ("v0", "v2")): torch.matmul,
    # frobenius inner product
    ((("v0", "v1"), ("v0", "v1")), ()): lambda A, B: torch.dot(A.flatten(), B.flatten()),
    ...
}
```

**Comment by nikitaved** [comment]
@randolf-scholz , regarding `torch.einsum` - the code there should be revisited, for sure. I am not sure that it follows the `opt_einsum` recommendations in full. `opt_einsum` explicitly tells whether to run `gemm/outer/dot`, etc...

**Comment by randolf-scholz** [comment]
@nikitaved `opt_einsum` solves a somewhat different problem, and for smaller tensors it can be slow due to overhead. I made a small demo of what I have in mind, here are some results:

```python
n = 1024
A = torch.randn(n, n).cpu()
B = torch.randn(n, n).cpu()

# bench trace on CPU
%timeit torch.trace(A)                  # 2.03 μs ± 90.3 ns per loop
%timeit torch.einsum("ii->", A)         # 7.69 μs ± 89.6 ns per loop
%timeit oe.contract("ii->", A)          # 29.6 μs ± 169 ns per loop 😕
%timeit my_einsum("ii->", A)            # 3.59 μs ± 44.1 ns per loop
# bench tensordot on CPU
%timeit torch.tensordot(A, B)           # 4.27 ms ± 54.2 μs per loop
%timeit torch.einsum("ij, ij->", A, B)  # 31.3 μs ± 215 ns per loop 😕
%timeit oe.contract("ij, ij->", A, B)   # 4.36 ms ± 42.8 μs per loop
%timeit my_einsum("ij, ij->", A, B)     # 21.6 μs ± 333 ns per loop 🤩

A = torch.randn(n, n).cuda()
B = torch.randn(n, n).cuda()

# bench trace on GPU
%timeit torch.trace(A)                  # 6.06 μs ± 80.1 ns per loop
%timeit torch.einsum("ii->", A)         # 11.9 μs ± 193 ns per loop
%timeit oe.contract("ii->", A)          # 34.7 μs ± 244 ns per loop 😕
%timeit my_einsum("ii->", A)            # 8.09 μs ± 88.5 ns per loop
# bench tensordot on GPU
%timeit torch.tensordot(A, B)           # 29 μs ± 19.6 ns per loop
%timeit torch.einsum("ij, ij->", A, B)  # 29.2 μs ± 26.1 ns per loop
%timeit oe.contract("ij, ij->", A, B)   # 39.5 μs ± 357 ns per loop
%timeit my_einsum("ij, ij->", A, B)     # 28.9 μs ± 31 ns per loop
```


<details><summary> Implementation </summary>

```python
from typing import Callable

import torch
from torch import Tensor
import opt_einsum as oe

type Arg = tuple[str, ...]
type Signature = tuple[tuple[Arg, ...], Arg]

KERNELS: dict[Signature, Callable[..., Tensor]] = {
    # sum
    ((("v0",),), ()): torch.sum,
    # vec-vec inner product
    ((("v0",), ("v0",)), ()): torch.dot,
    # vec-vec elementwise multiply
    ((("v0",), ("v0",)), ("v0",)): torch.multiply,
    # trace
    ((("v0", "v0"),), ()): torch.trace,
    # matmul A B
    ((("v0", "v1"), ("v1", "v2")), ("v0", "v2")): torch.matmul,
    # frobenius inner product
    ((("v0", "v1"), ("v0", "v1")), ()): lambda A, B: torch.dot(
        A.flatten(), B.flatten()
    ),
}


def parse_einsum(einsum_str: str) -> Signature:
    """
    Very simplistic parser that splits something like 'ij, jk -> ik'
    into ( (('i','j'), ('j','k')), ('i','k') ).
    In practice, you'll likely need a more robust parser.
    """
    # Split into 'ij, jk' vs 'ik'
    lhs, rhs = einsum_str.split("->")
    # Split the left-hand side by commas -> ['ij', ' jk']
    lhs_factors = lhs.split(",")
    lhs_factors = [f.strip() for f in lhs_factors]

    # For each factor, split into individual indices
    lhs_parsed = [tuple(factor) for factor in lhs_factors]

    # For the right-hand side
    rhs = rhs.strip()
    rhs_parsed = tuple(rhs)

    return canonicalize_indices(tuple(lhs_parsed), rhs_parsed)


class AutoIndexMap(dict[str, str]):
    def __missing__(self, key: str) -> str:
        new_label = f"v{len(self)}"
        self[key] = new_label
        return new_label


def canonicalize_indices(lhs: tuple[Arg, ...], rhs: Arg) -> Signature:
    index_map = AutoIndexMap()
    lhs_canonical = [tuple(index_map[idx] for idx in arg) for arg in lhs]
    rhs_canonical = tuple(index_map[idx] for idx in rhs)
    return tuple(lhs_canonical), rhs_canonical


def my_einsum(spec: str, *tensors: Tensor) -> Tensor:
    signature = parse_einsum(spec)
    kernel = KERNELS[signature]
    return kernel(*tensors)
```

</details>

**Comment by nikitaved** [comment]
@randolf-scholz , we cannot optimize kernel selection paths in eager mode. Have you tried one of them compilers?
